src/TpetraLinearSystem.C

```diff
@@ -798,16 +852,41 @@ void TpetraLinearSystem::fill_entity_to_row_LID_mapping()
 
 void TpetraLinearSystem::fill_entity_to_col_LID_mapping()
 {
-    const stk::mesh::BulkData& bulk = realm_.bulk_data();
+  const stk::mesh::BulkData& bulk = realm_.bulk_data();
+  stk::mesh::Selector selector = bulk.mesh_meta_data().universal_part() & !(realm_.get_inactive_selector());
     entityToColLID_ = LinSys::EntityToLIDView("entityToLID",bulk.get_size_of_entity_index_space());
-    const stk::mesh::BucketVector& nodeBuckets = bulk.buckets(stk::topology::NODE_RANK);
+    const stk::mesh::BucketVector& nodeBuckets = realm_.get_buckets(stk::topology::NODE_RANK,selector);
     for(const stk::mesh::Bucket* bptr : nodeBuckets) {
         const stk::mesh::Bucket& b = *bptr;
         const stk::mesh::EntityId* nodeIds = stk::mesh::field_data(*realm_.naluGlobalId_, b);
         for(size_t i=0; i<b.size(); ++i) {
             stk::mesh::Entity node = b[i];
-            GlobalOrdinal gid = GID_(nodeIds[i], numDof_, 0);
-            entityToColLID_[node.local_offset()] = totalColsMap_->getLocalElement(gid);
+
+            GlobalOrdinal gid =-1;
+            // needed because of some shared and ghost nodes. 
+            if (nodeIds[i] != bulk.identifier(node)) {
+              stk::mesh::Entity master = get_entity_master(bulk, node, nodeIds[i]);
+              if (master != node) {
+                gid = * stk::mesh::field_data(*realm_.tpetGlobalId_, master);
+                entityToColLID_[node.local_offset()] = totalColsMap_->getLocalElement(gid);
+              }
+              else
+                gid =  * stk::mesh::field_data(*realm_.tpetGlobalId_, node);
+            }
+            else
+              gid =  * stk::mesh::field_data(*realm_.tpetGlobalId_, node);
+
+            // GlobalOrdinal gid = tpetGIds[i];
+            
+            if(gid == 0 || gid == -1 || gid == std::numeric_limits<LinSys::GlobalOrdinal>::max() ) {
+              // This is needed because of a very small chance a ghost node doesn't have a master
+              // with a valid gid. ctests shows that entityToColLID_ is never accessed TpetraLinearSystem
```

Comment:
A very small chance? Can it happen or not?

Reply:
What is the situation where a ghosted node is not a `master` i.e., `(naluGlobalId_[node] != bulk.identifier(node)`?

Reply:
This happens only in ctest::unit_test1. In the case that a ghosted node does not have a master with a valid tpetGlobalID, it is never accessed downstream. 

This case should never happen, if my understanding of the STK documentation is correct, however unit_test1 does trigger this behavior. It may be that unit_test1 has a broken mesh set up.

I will modify this comment.

Reply:
It's puzzling that it could appear in unit_test1. Unit_test1 is the single-processor run of the unit-tests, and there is no ghosting when there is only 1 MPI rank.